axes/handlers/cache.py

```diff
@@ -0,0 +1,116 @@
+        cache = get_axes_cache()
+        cache_key = get_client_cache_key(request, credentials)
+        return cache.get(cache_key, default=0)
+
+    def user_login_failed(self, sender, credentials, request=None, **kwargs):  # pylint: disable=too-many-locals
+        """
+        When user login fails, save attempt record in cache and lock user out if necessary.
+
+        :raises AxesSignalPermissionDenied: if user should be locked out.
+        """
+
+        if request is None:
+            log.error('AXES: AxesCacheHandler.user_login_failed does not function without a request.')
+            return
+
+        username = get_client_username(request, credentials)
+        ip_address = get_client_ip_address(request)
+        user_agent = get_client_user_agent(request)
+        path_info = get_client_path_info(request)
+        client_str = get_client_str(username, ip_address, user_agent, path_info)
+
+        if self.is_whitelisted(request, credentials):
+            log.info('AXES: Login failed from whitelisted client %s.', client_str)
+            return
+
+        failures_since_start = 1 + self.get_failures(request, credentials)
+
+        if failures_since_start > 1:
+            log.warning(
+                'AXES: Repeated login failure by %s. Count = %d of %d. Updating existing record in the cache.',
+                client_str,
+                failures_since_start,
+                settings.AXES_FAILURE_LIMIT,
+            )
+        else:
+            log.warning(
+                'AXES: New login failure by %s. Creating new record in the cache.',
+                client_str,
+            )
+
+        cache = get_axes_cache()
+        cache_key = get_client_cache_key(request, credentials)
+        cache_timeout = get_cool_off().total_seconds()
+
+        cache.set(cache_key, failures_since_start, cache_timeout)
+
+        if failures_since_start >= settings.AXES_FAILURE_LIMIT:
+            log.warning('AXES: Locking out %s after repeated login failures.', client_str)
+
+            user_locked_out.send(
+                'axes',
+                request=request,
+                username=username,
+                ip_address=ip_address,
+            )
+
+            raise AxesSignalPermissionDenied('Locked out due to repeated login failures.')
+
+    def user_logged_in(self, sender, request, user, **kwargs):  # pylint: disable=unused-argument
+        """
+        When user logs in, update the AccessLog related to the user.
+        """
+
+        username = user.get_username()
+        credentials = get_credentials(username)
+        ip_address = get_client_ip_address(request)
+        user_agent = get_client_user_agent(request)
+        path_info = get_client_path_info(request)
+        client_str = get_client_str(username, ip_address, user_agent, path_info)
+
+        log.info('AXES: Successful login by %s.', client_str)
+
+        if settings.AXES_RESET_ON_SUCCESS:
+            cache = get_axes_cache()
```

Comment:
@aleksihakli aren't these lines the same as `self.get_failures`?

Reply:
Yeah, these methods still have some overlap and need some refactoring and cleanup. I'd just like to keep the delta between the readability of the database and cache backend small so they are easy to compare.